### DISCUSSION TITLE
ibek-support git submodule in Generic IOC not initialised during build

### DISCUSSION THREAD

**Discussion by ulrikpedersen** [epics-containers/epics-containers.github.io#114] [Q&A] [answered]
Just a little feedback to consider - I'm not sure if anything specifically is broken or if I've just done something wrong... This morning, I thought I've broken everything - I got myself wound into a knot trying to update the ibek-support submodule with a couple of very minor changes. I managed to make the relevant change and commit them to both my forked ibek-support repo and generic IOC repo. GH Actions now builds so repos are in a good state.

However, my local build was broken - there were issues with my added files from devcontainer belonging to root and things got messed up trying to fix that with chown. But because the repos were now in a building state on github, I decided to wipe my local ioc-ads directory entirely and clone it again. I followed the steps from the ["Changing Generic IOC"](https://epics-containers.github.io/main/tutorials/ioc_changes2.html#preparation) tutorial. Only now the `./build` fails - because the ibek-support git submodule isn't initialised (as the tutorial says it should). I'm thinking maybe the bash script line:  `if [ ! -d ibek-support ] ; then git submodule update --init ; fi` isn't what is required here? Because after cloning the generic IOC repo, the `ibek-support/` folder *does* exist - because it is a submodule. Perhaps the build script should instead check if the directory is empty?

Running the `git submodule update --init` manually and then running `./build` again works and the build completes without error.

I'm running bash 5.1.16 on Ubuntu 22.04.

**Comment by ulrikpedersen** [comment]
(this question is transferred from email thread with @gilesknap this morning...)

**Comment by gilesknap** [comment, marked as answer]
It is not intentional that the folder exists until you init the submodule but maybe an empty one has crept into the template. I'll check that. I do agree that a more robust check is required. I'm updating the template right now and will put that in.

Regarding use of docker. We are a podman shop and it works beautifully with devcontainers such that file ownership is never an issue. Nevertheless I do work with docker at home and have the following tips that may not yet be documented.

- most important is that you should set EC_REMOTE_USER to vscode.  Set that in your bashrc and forget about it. See https://github.com/epics-containers/ioc-pmac/blob/6c5880f909085767f4b78aca1c0a9e3bbe2e86c0/.devcontainer/devcontainer.json#L28-L32 and also https://github.com/epics-containers/ioc-pmac/blob/6c5880f909085767f4b78aca1c0a9e3bbe2e86c0/.devcontainer/postCreateCommand#L11-L15
- Once you have the above it is usually find to just do 
sudo chown -R vscode /root-of-troublesome-repo if things go wrong (but they should not ususaly)

**Reply by ulrikpedersen** [reply]
Thanks @gilesknap the EC_REMOTE_USER trick works for me 🎉

**Reply by ulrikpedersen** [reply]
Regarding the ibek-support git submodule. I don't think it's a mistake that the folder exists - I think that happens always when you have a submodule, even before its initialised.

For example look at the ioc-adsimdetector - it also has an empty ibek-support/ folder in it when freshly cloned:

```
ukp@osloxf04:~/EPICS$ git clone git@github.com:epics-containers/ioc-adsimdetector.git
Cloning into 'ioc-adsimdetector'...
remote: Enumerating objects: 432, done.
remote: Counting objects: 100% (281/281), done.
remote: Compressing objects: 100% (172/172), done.
remote: Total 432 (delta 130), reused 214 (delta 88), pack-reused 151
Receiving objects: 100% (432/432), 87.01 KiB | 492.00 KiB/s, done.
Resolving deltas: 100% (184/184), done.
ukp@osloxf04:~/EPICS$ cd ioc-adsimdetector/
ukp@osloxf04:~/EPICS/ioc-adsimdetector$ ls -alh
total 88K
drwxrwxr-x 10 ukp ukp 4.0K Mar 25 14:13 .
drwxrwxr-x  7 ukp ukp 4.0K Mar 25 14:13 ..
-rwxrwxr-x  1 ukp ukp  972 Mar 25 14:13 build
-rw-rw-r--  1 ukp ukp  308 Mar 25 14:13 .copier-answers.yml
drwxrwxr-x  2 ukp ukp 4.0K Mar 25 14:13 .devcontainer
-rw-rw-r--  1 ukp ukp 2.2K Mar 25 14:13 Dockerfile
-rw-rw-r--  1 ukp ukp   55 Mar 25 14:13 .dockerignore
drwxrwxr-x  8 ukp ukp 4.0K Mar 25 14:13 .git
drwxrwxr-x  3 ukp ukp 4.0K Mar 25 14:13 .github
-rw-rw-r--  1 ukp ukp  437 Mar 25 14:13 .gitignore
drwxrwxr-x  2 ukp ukp 4.0K Mar 25 14:13 .gitlab
-rw-rw-r--  1 ukp ukp  917 Mar 25 14:13 .gitlab-ci.yml
-rw-rw-r--  1 ukp ukp  108 Mar 25 14:13 .gitmodules
drwxrwxr-x  2 ukp ukp 4.0K Mar 25 14:13 ibek-support           <-- there it is
drwxrwxr-x  4 ukp ukp 4.0K Mar 25 14:13 ioc
-rw-rw-r--  1 ukp ukp  12K Mar 25 14:13 LICENSE
-rw-rw-r--  1 ukp ukp  382 Mar 25 14:13 README.md
-rw-rw-r--  1 ukp ukp  145 Mar 25 14:13 requirements.txt
drwxrwxr-x  3 ukp ukp 4.0K Mar 25 14:13 services
drwxrwxr-x  3 ukp ukp 4.0K Mar 25 14:13 tests
ukp@osloxf04:~/EPICS/ioc-adsimdetector$ ls -alh ibek-support/
total 8.0K
drwxrwxr-x  2 ukp ukp 4.0K Mar 25 14:13 .
drwxrwxr-x 10 ukp ukp 4.0K Mar 25 14:13 ..
ukp@osloxf04:~/EPICS/ioc-adsimdetector$ if [ ! -d ibek-support ]; then echo "git submodule init required"; fi
ukp@osloxf04:~/EPICS/ioc-adsimdetector$ 
```

I think the test should instead be if the ibek-support/ folder is empty...

**Reply by gilesknap** [reply]
OK I've updated the script to look for ibek-support/_globals and that should do the trick.
should be released no later than tomorrow.

**Reply by gilesknap** [reply]
This is released in ioc-template 3.4.0

**Reply by gilesknap** [reply]
BTW I'm going to write 'epics-containers Versioning Manifesto' which talks about how changes will be made to the framework without breaking existing work and how to then adopt those changes in a non-breaking fashion.
